src/gtk/toga_gtk/widgets/tree.py

```diff
@@ -1,41 +1,133 @@
+    # this method to access a private property.
+
+    # toga.sources.base.Row -> list (row's items, converted to string)
+    def _row_items(self, row):
+        return [str(getattr(row, attr)) for attr in self.interface._accessors]
+
+    # sources.base.Node -> int (node's index)
+    def _node_index(self, node):
+        def _parent(node):
+            return getattr(node, "_parent", None) or getattr(node, "_source", None)
+
+        for i, sibling in enumerate(_parent(node)):
+            if sibling is node:
+                return i
+
+    # sources.base.Node -> tuple (node's path)
+    def _node_path(self, node):
+        def f(node, path=[]):
+            if hasattr(node, "_parent"):
+                path.insert(0, self._node_index(node))
+                return f(node._parent, path)
+            else:
+                return path
+
+        return tuple(f(node))
+
+    # sources.base.Node -> Gtk.TreeIter
+    def _node_iter(self, node):
+        if hasattr(node, "_parent"):
+            return self.store.get_iter(self._node_path(node))
+        else:
+            return None
+
+    # tuple (node's path) -> sources.base.Node
+    def _get_node(self, path: tuple):
+        lpath = list(path)
+        lpath.reverse() # [level2, level1, level0]
+
+        def _children(node):
+            return getattr(node, "_children", None) or getattr(node, "_roots", None)
+
+        def f(node, path):
+            if _children(node) and lpath:
+                index = lpath.pop()
+                return f(node[index], lpath)
+            else:
+                return node
+
+        return f(self.interface.data, lpath)
+
+    def _on_select(self, selection):
+        if hasattr(self.interface, "_on_select") and self.interface.on_select:
+            tree_model, tree_iter = selection.get_selected()
+
+            if tree_iter:
+                tree_path = tree_model.get_path(tree_iter)
+                path = tree_path.get_indices()
+            else:
+                path = None
+
+            self.interface.on_select(None, node=self._get_node(path))
+
+    def _refresh(self):
+
+        # Temporarily disconnecting the TreeStore improves performance for large
+        # updates by deferring row rendering until the update is complete.
+        self.treeview.set_model(None)
+
+        self.store.clear()
+
+        def append_node(parent_node, it):
+            for i, child_node in enumerate(parent_node):
+                self.store.append(it, self._row_items(child_node))
+                it = self.store.iter_nth_child(it, i)
+                append_node(child_node, it)
+
+        append_node(self.interface.data, None)
+
+        self.treeview.set_model(self.store)
+
+    def change_source(self, source):
+        self._refresh()
+
+    def insert(self, item, index, **kwargs):
+        self.store.insert(
+            self._node_iter(item._parent),
+            index,
+            self._row_items(item)
+        )
```

Comment:
Your test case worked! However, I found some [old docs for pygtk2] (http://www.pygtk.org/pygtk2tutorial/sec-TreeModelInterface.html) stating that not all versions of Gtk are consistent with this behavior. You can check if a row iters are persisent with `treestore.get_flags()`, which will include the `TREE_MODEL_ITERS_PERSIST` flag. I suppose my helpers could be used a fallback when that flag is absent, but there is another option altogether, call Gtk.TreeRowReference. Those are always persistent, though slightly less performant. Another [source](http://scentric.net/tutorial/sec-treemodel-rowref.html) states that both `Gtk.ListStore` and `Gtk.TreeStore` use persistent iters, but that `Gtk.TreeRowReference` is the "Right Way". 
Sidenote: It always irks me when a problem has two solutions and the more performant one is also the discouraged one :)

In the future, it may be worth making some benchmarks or something to determine which of the solutions on the table is most suitable (currently, these are my helper functions, persistent Gtk.TreeIter, and Gtk.TreeRowReference).

Anyway, I don't mind refactoring a little to use `Gtk.TreeRowReference` as the Node `_impl`. Although my research shows that using the iter returned from `insert` should [theoretically] be fine, something in me sees it as a risk of unexpected behavior.

Unless you have further opinion on this matter, I'll go with Gtk.TreeRowReference since it's purpose is clearly for keeping track of rows as the data changes. If we come up with something better in the future we can adapt.

Reply:
I wouldn't be too worried about that documentation - pygtk2 is an ancient interface; when they talk about "older" versions of GTK, I wouldn't be surprised if they're referring to GTK2 - almost 10 years old at this point. The APIs for GTK2 are completely different across the board. 

So I'd suggest moving to TreeRowReference, and not worrying about the "older" behavior, or providing shims for the "old" API if persistent references aren't supported. If someone on a platform that is worth supporting reports issues, we can revisit this.